### ISSUE TITLE
greeter-compositor does not start on NixOS with mutter 45.rc

### ISSUE THREAD

**Issue by bobby285271** [elementary/greeter#682]
By copying the `vapi` changes from https://github.com/elementary/gala/pull/1746/commits/8b65beb1d60e51af26cfda2819e837d3e3975e39 (https://github.com/elementary/greeter/commit/4c38949d7f307e3c576bc6f7d240106a72c2902a) it already builds but it looks quite weird (as shown in screenshot, I do can use it to login / unlock screen though :woman_shrugging:), not sure if this can be reproduced somewhere.

~~There are no crash or failed units from my first observations.~~ ？

![屏幕截图 2023-08-29 20 56 23](https://github.com/elementary/greeter/assets/20080233/570b0fec-1009-4018-baf0-315af1f925f1)

**Comment by bobby285271** [comment]
This is `cat /var/log/lightdm/seat0-greeter.log`

<details>

```
[root@nixos:/home/test]# cat /var/log/lightdm/seat0-greeter.log 

(io.elementary.greeter:958): Handy-WARNING **: 13:57:03.958: Failed to load icon: Error opening file /home/test/.face: Permission denied

** (io.elementary.greeter:958): CRITICAL **: 13:57:04.226: MainWindow.vala:360: prompt: `Password: ' (1)

** (io.elementary.greeter:958): CRITICAL **: 13:57:04.639: MainWindow.vala:360: prompt: `Password: ' (1)

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.235: g_variant_new_string: assertion 'string != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.235: g_variant_ref_sink: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.235: g_variant_ref_sink: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-GIO-CRITICAL **: 13:57:05.235: g_settings_schema_key_type_check: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.235: g_variant_get_type_string: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-GIO-CRITICAL **: 13:57:05.236: g_settings_set_value: key 'document-font-name' in 'org.gnome.desktop.interface' expects type 's', but a GVariant of type '(null)' was given

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.236: g_variant_unref: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.244: g_variant_new_string: assertion 'string != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.244: g_variant_ref_sink: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.244: g_variant_ref_sink: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-GIO-CRITICAL **: 13:57:05.244: g_settings_schema_key_type_check: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.244: g_variant_get_type_string: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-GIO-CRITICAL **: 13:57:05.244: g_settings_set_value: key 'font-name' in 'org.gnome.desktop.interface' expects type 's', but a GVariant of type '(null)' was given

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.244: g_variant_unref: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.252: g_variant_new_string: assertion 'string != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.252: g_variant_ref_sink: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.252: g_variant_ref_sink: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-GIO-CRITICAL **: 13:57:05.252: g_settings_schema_key_type_check: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.252: g_variant_get_type_string: assertion 'value != NULL' failed

(io.elementary.greeter:958): GLib-GIO-CRITICAL **: 13:57:05.252: g_settings_set_value: key 'monospace-font-name' in 'org.gnome.desktop.interface' expects type 's', but a GVariant of type '(null)' was given

(io.elementary.greeter:958): GLib-CRITICAL **: 13:57:05.252: g_variant_unref: assertion 'value != NULL' failed

** (io.elementary.greeter:958): CRITICAL **: 13:57:06.522: SubprocessSupervisor.vala:44: Child process exited with code 1
```

</details>

**Comment by lenemter** [comment]
> SubprocessSupervisor.vala:44: Child process exited with code 1

Seems like greeter compositor crashes, which explains why there's black background.

**Comment by bobby285271** [comment]
Still an issue

I have no idea how to get backtrace here (as shown in the screenshot [above](https://github.com/elementary/greeter/issues/682#issue-1871627523) there is nothing is coredumpctl). Later I tried removing these lines and built another VM:

https://github.com/elementary/greeter/blob/dbd9b6f9701f5992c3b3257c025b9cd80d041cc8/src/Application.vala#L40-L44

And I tried starting `io.elementary.greeter-compositor` myself (I think I pasted all commands I ran here):

```
[root@nixos:~]# export XAUTHORITY=/var/lib/lightdm/.Xauthority 

[root@nixos:~]# export DISPLAY=:0

[root@nixos:~]# pgrep -fa greeter
957 /nix/store/7nrk4vxksyq39wb2pmfg1y7c841zcl7v-elementary-greeter-7.0.0/bin/io.elementary.greeter

[root@nixos:~]# for i in gala mutter compositor; do pgrep -fa $i; done

[root@nixos:~]# pushd /nix/store/7nrk4vxksyq39wb2pmfg1y7c841zcl7v-elementary-greeter-7.0.0/bin
/nix/store/7nrk4vxksyq39wb2pmfg1y7c841zcl7v-elementary-greeter-7.0.0/bin ~

[root@nixos:/nix/store/7nrk4vxksyq39wb2pmfg1y7c841zcl7v-elementary-greeter-7.0.0/bin]# ./io.elementary.greeter-compositor 
libmutter-Message: 16:45:03.136: Running Mutter(GreeterCompositor) (using mutter 45.rc) as a X11 window and compositing manager
(io.elementary.greeter-compositor:1320): GLib-GIO-DEBUG: 16:45:03.139: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
(io.elementary.greeter-compositor:1320): GLib-GIO-DEBUG: 16:45:03.144: _g_io_module_get_default: Found default implementation dconf (DConfSettingsBackend) for ‘gsettings-backend’
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.146: watch_fast: "/org/gnome/desktop/interface/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.146: watch_fast: "/org/gnome/desktop/privacy/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.146: watch_fast: "/org/gnome/mutter/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.146: watch_fast: "/org/gnome/mutter/wayland/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.152: watch_established: "/org/gnome/desktop/interface/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.152: watch_established: "/org/gnome/desktop/privacy/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.152: watch_established: "/org/gnome/mutter/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.153: watch_established: "/org/gnome/mutter/wayland/" (establishing: 1)
(io.elementary.greeter-compositor:1320): GLib-GIO-DEBUG: 16:45:03.302: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.313: watch_fast: "/org/gnome/settings-daemon/peripherals/touchscreen/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.314: watch_established: "/org/gnome/settings-daemon/peripherals/touchscreen/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.318: watch_fast: "/org/gnome/desktop/interface/" (establishing: 0, active: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.319: watch_fast: "/org/gnome/desktop/peripherals/mouse/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.321: watch_established: "/org/gnome/desktop/peripherals/mouse/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.321: watch_fast: "/org/gnome/desktop/a11y/mouse/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.321: watch_established: "/org/gnome/desktop/a11y/mouse/" (establishing: 1)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.390: Added axis 'Abs X' (min:0.00, max:65535.00d, res:0) of device 2
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.391: Added axis 'Abs Y' (min:0.00, max:65535.00d, res:0) of device 2
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.392: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 2
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.393: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 2
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Scroll valuator 2: horizontal, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Scroll valuator 3: vertical, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Created device 'Virtual core pointer' (id: 2, has-cursor: yes)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Created device 'Virtual core keyboard' (id: 3, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 4
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 4
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Created device 'Virtual core XTEST pointer' (id: 4, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.395: Created device 'Virtual core XTEST keyboard' (id: 5, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.411: Created device 'Power Button' (id: 6, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.417: Added axis 'Abs X' (min:0.00, max:65535.00d, res:0) of device 7
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.421: Added axis 'Abs Y' (min:0.00, max:65535.00d, res:0) of device 7
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.421: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 7
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.421: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 7
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.421: Scroll valuator 2: horizontal, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.421: Scroll valuator 3: vertical, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.421: Created device 'QEMU QEMU USB Tablet' (id: 7, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.433: Created device 'QEMU Virtio Keyboard' (id: 8, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.440: Created device 'AT Translated Set 2 keyboard' (id: 9, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.447: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 10
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.448: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 10
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.452: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 10
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.452: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 10
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.452: Scroll valuator 2: horizontal, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.452: Scroll valuator 3: vertical, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.452: Created device 'VirtualPS/2 VMware VMMouse' (id: 10, has-cursor: no)
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.464: Added axis 'Abs X' (min:0.00, max:65535.00d, res:0) of device 11
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.464: Added axis 'Abs Y' (min:0.00, max:65535.00d, res:0) of device 11
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.465: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 11
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.465: Added axis 'Ignored' (min:-1.00, max:-1.00d, res:0) of device 11
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.465: Scroll valuator 2: horizontal, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.465: Scroll valuator 3: vertical, increment: 120.000000
(io.elementary.greeter-compositor:1320): libmutter-DEBUG: 16:45:03.465: Created device 'VirtualPS/2 VMware VMMouse' (id: 11, has-cursor: no)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.475: watch_fast: "/org/gnome/desktop/peripherals/mouse/" (establishing: 0, active: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.475: watch_fast: "/org/gnome/desktop/peripherals/touchpad/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.475: watch_fast: "/org/gnome/desktop/peripherals/trackball/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.475: watch_fast: "/org/gnome/desktop/peripherals/keyboard/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.475: watch_fast: "/org/gnome/desktop/a11y/keyboard/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.475: watch_established: "/org/gnome/desktop/peripherals/touchpad/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.476: watch_established: "/org/gnome/desktop/peripherals/trackball/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.476: watch_established: "/org/gnome/desktop/peripherals/keyboard/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.476: watch_established: "/org/gnome/desktop/a11y/keyboard/" (establishing: 1)
(io.elementary.greeter-compositor:1320): GLib-GIO-DEBUG: 16:45:03.495: _g_io_module_get_default: Found default implementation local (GLocalVfs) for ‘gio-vfs’
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.509: watch_fast: "/org/gnome/desktop/wm/preferences/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.511: watch_established: "/org/gnome/desktop/wm/preferences/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.512: watch_fast: "/org/gnome/mutter/" (establishing: 0, active: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.512: watch_fast: "/org/gnome/desktop/peripherals/mouse/" (establishing: 0, active: 2)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.512: watch_fast: "/org/gnome/desktop/interface/" (establishing: 0, active: 2)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.512: watch_fast: "/org/gnome/desktop/input-sources/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.512: watch_established: "/org/gnome/desktop/input-sources/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.523: watch_fast: "/org/gnome/desktop/wm/keybindings/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.523: watch_fast: "/org/gnome/mutter/keybindings/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.524: watch_fast: "/org/gnome/mutter/wayland/keybindings/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.524: watch_established: "/org/gnome/desktop/wm/keybindings/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.524: watch_established: "/org/gnome/mutter/keybindings/" (establishing: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.524: watch_established: "/org/gnome/mutter/wayland/keybindings/" (establishing: 1)
(io.elementary.greeter-compositor:1320): GLib-DEBUG: 16:45:03.548: g_unix_open_pipe() called with FD_CLOEXEC; please migrate to using O_CLOEXEC instead
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.558: watch_fast: "/org/gnome/desktop/input-sources/" (establishing: 0, active: 1)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.596: watch_fast: "/org/gnome/desktop/interface/" (establishing: 0, active: 3)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.599: watch_fast: "/org/gnome/desktop/a11y/applications/" (establishing: 0, active: 0)
(io.elementary.greeter-compositor:1320): dconf-DEBUG: 16:45:03.599: watch_established: "/org/gnome/desktop/a11y/applications/" (establishing: 1)
libEGL warning: DRI2: failed to authenticate
(mutter-x11-frames:1332): GLib-GIO-DEBUG: 16:45:03.810: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
(mutter-x11-frames:1332): GLib-GIO-DEBUG: 16:45:03.830: _g_io_module_get_default: Found default implementation dconf (DConfSettingsBackend) for ‘gsettings-backend’
(mutter-x11-frames:1332): dconf-DEBUG: 16:45:03.832: watch_fast: "/org/gnome/desktop/interface/" (establishing: 0, active: 0)
(mutter-x11-frames:1332): GLib-GIO-DEBUG: 16:45:03.844: _g_io_module_get_default: Found default implementation local (GLocalVfs) for ‘gio-vfs’
(mutter-x11-frames:1332): GLib-GIO-DEBUG: 16:45:03.851: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
(mutter-x11-frames:1332): dconf-DEBUG: 16:45:03.854: watch_established: "/org/gnome/desktop/interface/" (establishing: 1)
Window manager error: Another compositing manager is already running on screen 0 on display “:0”.

[root@nixos:/nix/store/7nrk4vxksyq39wb2pmfg1y7c841zcl7v-elementary-greeter-7.0.0/bin]# echo $?
1
```

I was quite confused here since the exit code matches the one in `/var/log/lightdm/seat0-greeter.log` and I did some `pgrep` ahead and did not found obvious things, later I tried `--replace` and the result is still the same?

```
[root@nixos:/nix/store/7nrk4vxksyq39wb2pmfg1y7c841zcl7v-elementary-greeter-7.0.0/bin]# unset G_MESSAGES_DEBUG

[root@nixos:/nix/store/7nrk4vxksyq39wb2pmfg1y7c841zcl7v-elementary-greeter-7.0.0/bin]# ./io.elementary.greeter-compositor --replace
libmutter-Message: 16:45:34.248: Running Mutter(GreeterCompositor) (using mutter 45.rc) as a X11 window and compositing manager
libEGL warning: DRI2: failed to authenticate
Window manager error: Another compositing manager is already running on screen 0 on display “:0”.
```

Later I tried building VM on nixos-unstable (which is on GNOME 44 and we pin Mutter on 43 for Pantheon, and greeter work fine there), I did not remove the lines in `greeter/src/Application.vala` but I killed `io.elementary.greeter-compositor` manually and using the above steps I was able to start it fine so I think I should be testing it right? :confused:

**Comment by bobby285271** [comment]
I finally got around to revisit this. The issue here is exactly [this one](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2445#note_1745867), and [this workaround](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2445#note_1747876) works at least for this issue.

*(I am still pinning mutter to a [patched 43.8](https://github.com/NixOS/nixpkgs/blob/8084770f0800fa7f8efd2106e8f567ca696a74d7/pkgs/desktops/gnome/core/mutter/43/default.nix) for NixOS Pantheon right now)*
